File: python/0144二叉树前序遍历.py
from binarytree import pre_create,TreeNode

# 递归
def preorderTraversal1(root: TreeNode) -> list:
    res=[]
    def _pot(r:TreeNode):
        
        if(r):
            res.append(r.val)
            _pot(r.left)
            _pot(r.right)
    
    _pot(root)
    
    return res

# 迭代
def preorderTraversal2(root: TreeNode):
    res=[]
    if(not root):
        return res
    
    st=[root]

    while(st):
        top=st.pop()
        res.append(top.val)

        if(top.right):
            st.append(top.right)
        if(top.left):
            st.append(top.left)
    
    return res


print(preorderTraversal2(pre_create([1,None,2,3])))

File: cpp/0144二叉树前序遍历.cpp
#include"mybitree.h"
#include<vector>
#include<stack>

// 递归
void Traversal1(TreeNode *root,vector<int>& res)
{
    if(root)
    {
        res.push_back(root->val);
        Traversal1(root->left, res);
        Traversal1(root->right, res);
    }
}

void Traversal2(TreeNode* root,vector<int>&res)
{
    stack<TreeNode*> st;
    TreeNode *temp;
    st.push(root);

    while(!st.empty())
    {
        temp = st.top();
        res.push_back(temp->val);
        st.pop();
        if (temp->right!=nullptr)
            st.push(temp->right);
        if(temp->left!=nullptr)
            st.push(temp->left);
    }

}

vector<int> preorderTraversal(TreeNode *root)
{
    vector<int> res;
    if(root==nullptr)
        return res;
    Traversal2(root, res);

    return res;
}

int main()
{
    deque<int> dq = {1,NULL, 2, 3};
    TreeNode *root = PreCreate(dq);

    vector<int> res = preorderTraversal(root);

    for(int i : res)
    {
        cout << i;
    }
}